kernel/src/transaction.rs

```diff
@@ -0,0 +1,265 @@
+use std::sync::Arc;
+
+use crate::actions::get_log_schema;
+use crate::path::ParsedLogPath;
+use crate::schema::{Schema, SchemaRef};
+use crate::snapshot::Snapshot;
+use crate::{DataType, Expression};
+use crate::{DeltaResult, Engine, EngineData};
+
+const KERNEL_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// A transaction represents an in-progress write to a table.
+pub struct Transaction {
+    read_snapshot: Arc<Snapshot>,
+    commit_info: Option<EngineCommitInfo>,
+}
+
+// Since the engine can include any commit info it likes, we unify the data/schema pair as a single
+// struct with Arc semantics.
+#[derive(Clone)]
+struct EngineCommitInfo {
+    data: Arc<dyn EngineData>,
+    schema: SchemaRef,
+}
+
+impl std::fmt::Debug for Transaction {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!(
+            "Transaction {{ read_snapshot version: {}, commit_info: {} }}",
+            self.read_snapshot.version(),
+            self.commit_info.is_some()
+        ))
+    }
+}
+
+impl Transaction {
+    /// Create a new transaction from a snapshot. The snapshot will be used to read the current
+    /// state of the table (e.g. to read the current version).
+    ///
+    /// Instead of using this API, the more typical (user-facing) API is
+    /// [Table::new_transaction](crate::table::Table::new_transaction) to create a transaction from
+    /// a table automatically backed by the latest snapshot.
+    pub(crate) fn new(snapshot: impl Into<Arc<Snapshot>>) -> Self {
+        Transaction {
+            read_snapshot: snapshot.into(),
+            commit_info: None,
+        }
+    }
+
+    /// Consume the transaction and commit the in-progress write to the table.
+    pub fn commit(self, engine: &dyn Engine) -> DeltaResult<CommitResult> {
+        // step one: construct the iterator of actions we want to commit
+        // note: only support commit_info right now.
+        let (actions, _actions_schema): (
+            Box<dyn Iterator<Item = Box<dyn EngineData>> + Send>,
+            SchemaRef,
+        ) = match self.commit_info {
+            Some(ref engine_commit_info) => {
+                let (actions, schema) = generate_commit_info(engine, engine_commit_info)?;
+                (Box::new(std::iter::once(actions)), schema)
+            }
+            None => {
+                (
+                    // if there is no commit info, actions are empty iterator and schema is just our
+                    // known log schema.
+                    Box::new(std::iter::empty()),
```

Comment:
Ah, this is why Box was needed... the iterator's type is not stable across match arms

Reply:
fixed now not an issue :)